File: lib/java/src/main/java/io/github/msgpact/internal/validation/ValidateUnionStruct.java
package io.github.msgpact.internal.validation;

import static io.github.msgpact.internal.validation.ValidateStructFields.validateStructFields;

import java.util.List;
import java.util.Map;

import io.github.msgpact.internal.types.VStruct;

public class ValidateUnionStruct {
    static List<ValidationFailure> validateUnionStruct(
            VStruct unionStruct,
            String unionTag,
            Map<String, Object> actual, Map<String, Object> selectedTags, ValidateContext ctx) {
        final var selectedFields = selectedTags == null ? null : (List<String>) selectedTags.get(unionTag);
        return validateStructFields(unionStruct.fields, selectedFields, actual, ctx);
    }
}

File: lib/py/msgpact/internal/validation/ValidateUnionStruct.py
from typing import TYPE_CHECKING, cast
from ...internal.validation.ValidationFailure import ValidationFailure

if TYPE_CHECKING:
    from ...internal.validation.ValidateContext import ValidateContext
    from ..types.VStruct import VStruct
    from ..types.VTypeDeclaration import VTypeDeclaration


def validate_union_struct(
    union_struct: 'VStruct',
    union_tag: str,
    actual: dict[str, object],
    selected_tags: dict[str, object],
    ctx: 'ValidateContext'
) -> list['ValidationFailure']:
    selected_fields = cast(list[str], selected_tags.get(
        union_tag)) if selected_tags else None
    from ...internal.validation.ValidateStructFields import validate_struct_fields

    return validate_struct_fields(
        union_struct.fields,
        selected_fields,
        actual,
        ctx
    )
